[FIX] account: Adding company default tax for bill and invoice.

daniel-stenlov opened the pull request:
Company default tax was removed from odoo 17 when the code was rewritten. This feature was available in Odoo 16. This code reimplements this feature for Odoo 17 if no tax is set on bill or invoice. This will also give a default tax on bills with only text lines.

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/169041.png)](https://mergebot.odoo.com/odoo/odoo/pull/169041)

svfu-odoo commented:
Thanks for the PR / suggestion.

It seems like the feature you want to (re)add was intentionally removed in commit 15b2c690e13b73e7e83e6b907b9e199c38c7382c

The default sale / purchase taxes are (still) used to set a tax on new products.
This is also what the tooltip of these taxes / settings suggest.
From the products the taxes are still applied to the invoice lines.

The behavior was changed because the default sale / purchase taxes were wrong often enough to create friction.

svfu-odoo reviewed (commented):
Thanks for the PR / suggestion.

It seems like the feature you want to (re)add was intentionally removed in commit 15b2c690e13b73e7e83e6b907b9e199c38c7382c

The default sale / purchase taxes are (still) used to set a tax on new products.
This is also what the tooltip of these taxes / settings suggest.
From the products the taxes are still applied to the invoice lines.

The behavior was changed because the default sale / purchase taxes were wrong often enough to create friction.

daniel-stenlov commented:
This is very interesting. I have read the pull request and I don't really understand why it was removed. Since there is just a statement that it should be removed and not why. My guess would be that it is connected to country specific issues. For example, In Sweden it create more work for a missing VAT, than it does with wrong VAT. So, any VAT is better than no VAT. It's also easy to miss to add VAT to a product if you don't use a system product for a BILL for example. You might just want to add a line and not a product and when you do it's easy to miss the VAT. So I do have two suggestions to solve this:

**Company Setting**
Add a setting on a company to add default sales/purchase tax or not. This will solve the problem when people have issues with it.

**Accounting Verify Rules**
Another way is to add a rule set to accounts for warnings. This way could be used to pop up a warning when you try to push create the invoice/bill. For example Sweden uses 4 different VAT 25, 12, 6 and 0%. They are used for different things and is common in different business. It would be nice to add a non-blocking warning for an account so for example Account 4000 needs to have 25, 12, 6 or 0% or it will pop up a warning. But I do think the changes to implement this would be bigger, however it would be a nice feature.

chklop commented:
No, we don't want that. You can add default VAT on products, you can even set default VAT on accounts (it's usually a bad idea but in some cases, for specific income/expense accounts, it could make sense). But we do not want "default VAT on empty lines" where no information is present.

We've seen through feedbacks that no, any VAT is better than no VAT, because bad VAT is bad VAT all the time. Instead of enforcing a default value when we have no information at all to propose is something that we purposefully removed, and it will remain that way.

At the end, default taxes are meant to be applied on products and accounts. And for the encoding of Vendor Bills (where using products is usually not as important/enforced), we have Account & Product prediction that will help the user encode the appropriate VAT.

daniel-stenlov commented:
@chklop

Thank you for your feedback. I understand your perspective, but I have to respectfully disagree. Different countries have different requirements, and there are varying needs that must be considered. While your statement may be correct in some cases, it does not apply universally. Flexibility is essential to meet both goals.

I appreciate your viewpoint and will retract the pull request to find an alternative solution that addresses these concerns better. The end customer has been informed about the situation and the reasons behind Odoo S.A.'s decision to remove the feature from Odoo 17.
